Tools/AP_Periph/can.cpp

```diff
@@ -727,643 +597,37 @@ void AP_Periph_FW::can_safety_button_update(void)
-#if DEBUG_PKTS
-    if (res < 0) {
-        can_printf("Tx error %d\n", res);
-    }
-#endif
-#if HAL_ENABLE_SENDING_STATS
-    if (res <= 0) {
-        protocol_stats.tx_errors++;
-    } else {
-        protocol_stats.tx_frames += res;
-    }
+    NVIC_SystemReset();
 #endif
-    return res > 0;
-}
-
-void AP_Periph_FW::processTx(void)
-{
-    for (CanardCANFrame* txf = NULL; (txf = canardPeekTxQueue(&dronecan.canard)) != NULL;) {
-        AP_HAL::CANFrame txmsg {};
-        txmsg.dlc = AP_HAL::CANFrame::dataLengthToDlc(txf->data_len);
-        memcpy(txmsg.data, txf->data, txf->data_len);
-        txmsg.id = (txf->id | AP_HAL::CANFrame::FlagEFF);
-#if HAL_CANFD_SUPPORTED
-        txmsg.canfd = txf->canfd;
-#endif
-        // push message with 1s timeout
-        bool sent = true;
-        const uint64_t now_us = AP_HAL::micros64();
-        const uint64_t deadline = now_us + 1000000U;
-        // try sending to all interfaces
-        for (auto &_ins : instances) {
-            if (_ins.iface == NULL) {
-                continue;
-            }
-    #if CANARD_MULTI_IFACE
-            if (!(txf->iface_mask & (1U<<_ins.index))) {
-                continue;
-            }
-    #endif
-    #if HAL_NUM_CAN_IFACES >= 2
-            if (can_protocol_cached[_ins.index] != AP_CAN::Protocol::DroneCAN) {
-                continue;
-            }
-    #endif
-            if (_ins.iface->send(txmsg, deadline, 0) <= 0) {
-                /*
-                  We were not able to queue the frame for
-                  sending. Only mark the send as failing if the
-                  interface is active. We consider an interface as
-                  active if it has had a successful transmit in the
-                  last 2 seconds
-                 */
-                volatile const auto *stats = _ins.iface->get_statistics();
-                uint64_t last_transmit_us = stats->last_transmit_us;
-                if (stats == nullptr || AP_HAL::micros64() - last_transmit_us < 2000000UL) {
-                    sent = false;
-                }
-            } else {
-#if CANARD_MULTI_IFACE
-                txf->iface_mask &= ~(1U<<_ins.index);
-#endif
-            }
-        }
-        if (sent) {
-            canardPopTxQueue(&dronecan.canard);
-            dronecan.tx_fail_count = 0;
-        } else {
-            // exit and try again later. If we fail 8 times in a row
-            // then cleanup any stale transfers to keep the queue from
-            // filling
-            if (dronecan.tx_fail_count < 8) {
-                dronecan.tx_fail_count++;
-            } else {
-#if HAL_ENABLE_SENDING_STATS
-                protocol_stats.tx_errors++;
-#endif
-                dronecan.tx_fail_count = 0;
-                cleanup_stale_transactions(now_us);
-            }
-            break;
-        }
-    }
-}
-
-#if HAL_ENABLE_SENDING_STATS
-void AP_Periph_FW::update_rx_protocol_stats(int16_t res)
-{
-    switch (res) {
-    case CANARD_OK:
```

Comment:
has this been lost or is it in canard library?

Reply:
its in AP_Canard_Iface already